pkg/apimachinery/patch/patch.go

```diff
@@ -38,6 +38,91 @@ const (
 	PatchRemoveOp  = "remove"
 )
 
+func (p *PatchOperation) MarshalJSON() ([]byte, error) {
+	switch p.Op {
+	// The 'remove' operation is the only patching operation without a value
+	// and it needs to be parsed differently.
+	case PatchRemoveOp:
+		return json.Marshal(&struct {
+			Op   string `json:"op"`
+			Path string `json:"path"`
+		}{
+			Op:   p.Op,
+			Path: p.Path,
+		})
+	case PatchTestOp, PatchReplaceOp, PatchAddOp:
+		return json.Marshal(&struct {
+			Op    string      `json:"op"`
+			Path  string      `json:"path"`
+			Value interface{} `json:"value"`
+		}{
+			Op:    p.Op,
+			Path:  p.Path,
+			Value: p.Value,
+		})
+	default:
+		return nil, fmt.Errorf("operation %s not recognized", p.Op)
+	}
+}
+
+type PatchSet struct {
+	patches []PatchOperation
+}
+
+type PatchOption func(patches *PatchSet)
+
+func New(opts ...PatchOption) *PatchSet {
+	p := &PatchSet{}
+	p.AddOption(opts...)
+	return p
+}
+
+func (p *PatchSet) AddOption(opts ...PatchOption) {
+	for _, f := range opts {
+		f(p)
+	}
+}
+
+func (p *PatchSet) addOp(op, path string, value interface{}) {
+	p.patches = append(p.patches, PatchOperation{
+		Op:    op,
+		Path:  path,
+		Value: value,
+	})
+}
+
+func WithTest(path string, value interface{}) PatchOption {
+	return func(p *PatchSet) {
+		p.addOp(PatchTestOp, path, value)
+	}
+}
+
+func WithAdd(path string, value interface{}) PatchOption {
+	return func(p *PatchSet) {
+		p.addOp(PatchAddOp, path, value)
+	}
+}
+
+func WithReplace(path string, value interface{}) PatchOption {
```

Comment:
As test is almost always used with replace, do you think it's a good idea to add a `WithTestAndReaplace` method?

Reply:
The idea of this PR was to avoid such combinations. Youu are free to combine all the provided basic options.

Reply:
> The idea of this PR was to avoid such combinations.

I don't see this idea being expressed anywhere on the PR or commit descriptions. Can you explain how did you come to this conclusion?

In any case, I'd be happy if you can provide a single example in our codebase that the test operation is being used without a replace coming right after it. It is at least very very rare if exists at all. I'm not even sure if there's a reason to use the test operation other than updates.

If you disagree, please explain what's the value of keeping those separated. As I see it, in practice, the test operation is always (or at least almost always) just a safer pre-step for updates.

Reply:
@iholder101 if you look at @EdDev comments, we should also probably always use add instead of test/replace

Reply:
> I don't see this idea being expressed anywhere on the PR or commit descriptions. Can you explain how did you come to this conclusion?

It is considered a good practice to do one thing and let callers combine options and not the other way around. It is a smell when you need to name something that does two things and I would suggest to avoid it.

> In any case, I'd be happy if you can provide a single example in our codebase that the test operation is being used without a replace coming right after it. It is at least very very rare if exists at all. I'm not even sure there's a reason to use the test operation other than updates.

This is surely controversial.
The use of the test is sometime used incorrectly, unfortunately taken for granted.
But even if it would be used always with a test, it is unreasonable to create a combination of all the operations with a test and consider that a good API.

> If you disagree, please explain what's the value of keeping those separated. As I see it, in practice, the test operation is always (or at least almost always) just a safer pre-step for updates.

As mentioned earlier, it was discussed in a different thread and reasoned for.

This solution is very elegant and clean, I hope to see more of this kind in the project.

Reply:
> But even if it would be used always with a test, it is unreasonable to create a combination of all the operations with a test and consider that a good API.

How have you come to the conclusion that I suggest to "create a combination of all the operations with a test"? That's really not what I've said.

An API that forces the user into duplicating obvious boiler-plate without any value is a bad API.

> if you look at @EdDev's comments, we should also probably always use add instead of test/replace

Thank you @alicefr. This is a good argument.
However, if that's true, please remove the usage of `WithTest()` from [here](https://github.com/kubevirt/kubevirt/pull/11599/files#diff-5e2591267ff7e9303f97ccac213986c2157131e3487a04a122b7662d401f7a57R233) or please explain why in this specific example it is necessary.

Reply:
I'm with Ed here.

If you look at the other PR, we ended up with a series of methods that combined different operations. I did it because of the reviews, but I wasn't particularly happy with them.
If we offer the single methods with `Test/rReplace` and then `TestAndReplace`, the user still needs to know which combinations they need to use. The only thing that can save is the need to repeat the path twice, but I like the fact that it is clear that those are 2 operations and not a single one.

@iholder101 does it sound reasonable to you?

Reply:
@alicefr I think we wrote a comment at the same time :) please look at https://github.com/kubevirt/kubevirt/pull/11599#discussion_r1549017640

Reply:
BTW, now that I think about it more, maybe we'd want to remove `WithTest()` altogether if it's never used in-code and we can't find a proper use-case for it. We can always add it later if needed.

Reply:
> I think in this case, it is needed.

Can you please elaborate?

Reply:
> How have you come to the conclusion that I suggest to "create a combination of all the operations with a test"? That's really not what I've said.
> 
> An API that forces the user into duplicating obvious boiler-plate without any value is a bad API.

Because that was discussed already and the state at that time was exactly that.
The suggestion to use `replace` with `test` implies that all the other combinations need to be introduced as well. At least, this is how my brain computes it, due to consistency and the wish for everything to look and feel the same.

Duplication is valid if it is expressive.
Instead of having `RunFooWithBoo()` we use `Run(WithFoo(), WithBoo())` because it constructs well the small pieces into one. It helps with testing and is a fit with the option to construct whatever options you want without creating a static combination.

We have examples of this already with the `go-client`, `libvmi` and more.

> > if you look at @EdDev's comments, we should also probably always use add instead of test/replace
> 
> Thank you @alicefr. This is a good argument. However, if that's true, please remove the usage of `WithTest()` from [here](https://github.com/kubevirt/kubevirt/pull/11599/files#diff-5e2591267ff7e9303f97ccac213986c2157131e3487a04a122b7662d401f7a57R233) or please explain why in this specific example it is necessary.

As I see it, the change there did not tried to fix the `add` instead of `replace`, so I am not sure why it needs to do it. A dedicated commit should say it changes it with a reason.

Reply:
> Instead of having RunFooWithBoo() we use Run(WithFoo(), WithBoo()) 

This is all nice and fine unless foo() and boo() are **always** (or 99% of the times) used together. If that's the case, separating them is creating a practically bad API because of theoretical principles that don't always fit reality. That's at least my opinion.

Reply:
> BTW, now that I think about it more, maybe we'd want to remove `WithTest()` altogether if it's never used in-code and we can't find a proper use-case for it. We can always add it later if needed.

`test` is needed is several cases, here are the ones I can think of:
- Changing a list/slice.
- Changing a complete subtree (e.g. the whole `Spec`).
- Changing a value that is dependent on the existence of another value in another place.

It is not needed when:
- Need to change one or more independent values in an atomic manner. If someone else changes it 1ns before or 1ns after, using a `test` will not help.